src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java

```diff
@@ -0,0 +1,512 @@
+                                  bucketResource),
+                              fileLoadModel.getSourcePath(),
+                              fileLoadModel.getTargetPath(),
+                              null);
+                        } catch (GoogleInternalServerErrorException e) {
+                          TimeUnit.SECONDS.sleep(5);
+                          attemptsLeft--;
+                          if (attemptsLeft == 0) {
+                            throw e;
+                          }
+                        }
+                      }
+                    } catch (Exception e) {
+                      logger.warn("Error copying file", e);
+                      return new CopyResult(
+                          null, fileLoadModel.getSourcePath(), fileLoadModel.getTargetPath(), e);
+                    }
+                  }));
+        }
+      }
+      List<CopyResult> results = FutureUtils.waitFor(futureFsFileInfos);
+      copyResults.addAll(results);
+    }
+
+    long succeededFiles = copyResults.stream().filter(f -> f.error() == null).count();
+
+    resultModel
+        .totalFiles(loadModels.size())
+        .succeededFiles((int) succeededFiles)
+        .failedFiles(loadModels.size() - (int) succeededFiles);
+
+    // Validate that the amount files copied / linked didn't exceed the maximum allowed specified
+    // by the user
+    validateErrors(workingMap, copyResults.stream().filter(r -> r.error() != null).toList());
+
+    // Save the results of the ingestion to be used to load into the load history table
+    List<BulkLoadHistoryModel> loadHistoryModels =
+        new ArrayList<>(copyResults.stream().map(this::fromCopyResult).toList());
+    workingMap.put(IngestMapKeys.BULK_LOAD_HISTORY_RESULT, loadHistoryModels);
+
+    loadedFiles.addAll(loadHistoryModels.stream().map(this::fromHistoryModel).toList());
+    return copyResults;
+  }
+
+  /**
+   * This method ensures that all filesystem metadata (e.g. data resulting from the target path in
+   * the file ingest) gets properly added to the dataset's firestore db
+   *
+   * @param loadModels The models being ingested whose directories and file entries must be created
+   *     Note: in the case of predictable file ids, this will be the subset of file models specified
+   *     by the user that were able to be successfully copied
+   * @param fileIdsByPath A lookup index that maps the virtual path of a file (e.g. what is being
+   *     recorded by this method) to the id for this file. If a conflict is found (e.g. from a
+   *     previous load) this map gets mutated and updated with the existing file id
+   */
+  @VisibleForTesting
+  void doCreateFileSystemEntries(
+      List<BulkLoadFileModel> loadModels, Map<String, UUID> fileIdsByPath)
+      throws InterruptedException {
+    logger.info("Extracting Directories");
+    // Get all directories to create:
+    List<String> directories =
```

Comment:
Continuing from my previous comment - could we build the list of FireStoreDirectoryEntries here and pass that to fileDao.upsertDirectoryEntries()?